TombEngine/Objects/TR4/Trap/SpikyWall.cpp

```diff
@@ -23,37 +29,88 @@ namespace TEN::Entities::Traps
 		auto& item = g_Level.Items[itemNumber];
 
 		item.ItemFlags[0] = item.TriggerFlags;
+		item.ItemFlags[1] = 0;
 	}
 
 	void ControlSpikyWall(short itemNumber)
 	{
 		auto& item = g_Level.Items[itemNumber];
 
-		if (!TriggerActive(&item) || item.Status == ITEM_DEACTIVATED)
+		if (!TriggerActive(&item))
 			return;
 
+		item.ItemFlags[0] = item.TriggerFlags;
+		
+		if (item.ItemFlags[1] && item.ItemFlags[0])
+			item.ItemFlags[1] -= 1;
+
 		int forwardVel = item.ItemFlags[0];
 		auto bounds = GameBoundingBox(&item);
 
-		// Get point collision.
-		auto pointColl = GetCollision(&item, item.Pose.Orientation.y, (forwardVel >= 0) ? bounds.Z2 : bounds.Z1);
-		int upperFloorBound = item.Pose.Position.y;
-		int lowerCeilBound = item.Pose.Position.y + bounds.Y1;
+		auto pointColl0 = GetCollision(&item, item.Pose.Orientation.y, (forwardVel >= 0) ? bounds.Z2 : bounds.Z1, bounds.Y2);
+		auto pointColl1 = GetCollision(&item, item.Pose.Orientation.y, (forwardVel >= 0) ? bounds.Z2 : bounds.Z1, bounds.Y2, (bounds.X2 - bounds.X1) / 2);
+
+		if (GetCollidedObjects(&item, CLICK(1), true, CollidedItems, CollidedMeshes, true))
+		{
+			int collidedItemNumber = 0;
+			while (CollidedItems[collidedItemNumber] != nullptr)
+			{
+				if (Objects[CollidedItems[collidedItemNumber]->ObjectNumber].intelligent)
+				{
+					CollidedItems[collidedItemNumber]->HitPoints = 0;
+				}
+				else if (CollidedItems[collidedItemNumber]->ObjectNumber == ID_SPIKY_WALL && !item.ItemFlags[1])
+				{
+					CollidedItems[collidedItemNumber]->TriggerFlags = 0;
+
+					item.TriggerFlags = 0;
+					item.Status = ITEM_DEACTIVATED;
+					item.ItemFlags[1] = 10;
+
+					StopSoundEffect(SFX_TR4_ROLLING_BALL);
+				}
+
+				collidedItemNumber++;
+			}
+		}
 
 		// Stop moving.
-		if (pointColl.Position.Floor < upperFloorBound ||
-			pointColl.Position.Ceiling > lowerCeilBound)
+		if (!item.TriggerFlags ||
+			pointColl0.Block->IsWall(item.Pose.Position.x, item.Pose.Position.z) || 
+			pointColl0.Block->Stopper ||
+			pointColl1.Block->IsWall(item.Pose.Position.x, item.Pose.Position.z) ||
+			pointColl1.Block->Stopper)
 		{
+			auto& room = g_Level.Rooms[item.RoomNumber];
+			for (auto& mesh : room.mesh)
+			{
+				if ((abs(pointColl0.Coordinates.x - mesh.pos.Position.x) < BLOCK(1) &&
+					abs(pointColl0.Coordinates.z - mesh.pos.Position.z) < BLOCK(1)) ||
+					abs(pointColl1.Coordinates.x - mesh.pos.Position.x) < BLOCK(1) &&
+					abs(pointColl1.Coordinates.z - mesh.pos.Position.z) < BLOCK(1) &&
+					StaticObjects[mesh.staticNumber].shatterType != ShatterType::None)
+				{					
+					if (mesh.HitPoints != 0)
+						continue;
+
+					mesh.HitPoints -= 1;
+					ShatterObject(nullptr, &mesh, -64, LaraItem->RoomNumber, 0);
+					SoundEffect(SFX_TR4_SMASH_ROCK, &item.Pose);
+					TestTriggers(item.Pose.Position.x, item.Pose.Position.y, item.Pose.Position.z, item.RoomNumber, true);
```

Comment:
In the future, we should make a function in the shatters objects that includes this "shattering" routine, so when any object wants to break them, (Lara shooting, rollingballs, spikewalls, enemies, etc) the objects always do it by their same routine. 

I'm concerned specially for the sound constant value, if at some point we neee to change it so each shatter have uses own sound, (rock, glass, wood, etc), having that in a single routine should be easier.

But as I see it, that's a task different of the purpose of this branch, so is is fine as it's, we better cover that in a different branch in the future, (as it may need to edit other objects as well)